tasks/files/snapshot.py

```diff
@@ -0,0 +1,871 @@
+        return "lvm bug encountered: %s" % " ".join(self.args)
+
+
+class SnapshotCommand:
+    SNAPSHOT = "snapshot"
+    SNAPSHOT_CHECK = "check"
+    SNAPSHOT_CLEAN = "clean"
+
+
+class SnapshotStatus:
+    SNAPSHOT_OK = 0
+    ERROR_INSUFFICENT_SPACE = 1
+    ERROR_ALREADY_DONE = 2
+    ERROR_SNAPSHOT_FAILED = 3
+    ERROR_REMOVE_FAILED = 4
+    ERROR_REMOVE_FAILED_NOT_SNAPSHOT = 5
+    ERROR_LVS_FAILED = 6
+    ERROR_NAME_TOO_LONG = 7
+    ERROR_ALREADY_EXISTS = 8
+    ERROR_NAME_CONFLICT = 9
+    ERROR_VG_NOTFOUND = 10
+    ERROR_LV_NOTFOUND = 11
+    ERROR_VERIFY_NOTSNAPSHOT = 12
+    ERROR_VERIFY_COMMAND_FAILED = 13
+    ERROR_VERIFY_NOT_FOUND = 14,
+    ERROR_CMD_INVALID = 15
+    ERROR_VERIFY_REMOVE_FAILED = 16,
+    ERROR_VERIFY_REMOVE_SOURCE_SNAPSHOT = 17
+
+
+# what percentage is part of whole
+def percentage(part, whole):
+    return 100 * float(part) / float(whole)
+
+
+# what is number is percent of whole
+def percentof(percent, whole):
+    return float(whole) / 100 * float(percent)
+
+
+def set_up_logging(log_dir="/tmp", log_prefix="snapshot_role"):
+    logger.setLevel(logging.DEBUG)
+
+    def make_handler(path, prefix, level):
+        log_file = "%s/%s.log" % (path, prefix)
+        log_file = os.path.realpath(log_file)
+        handler = logging.FileHandler(log_file)
+        handler.setLevel(level)
+        formatter = logging.Formatter(
+            "%(asctime)s %(levelname)s %(name)s/%(threadName)s: %(message)s"
+        )
+        handler.setFormatter(formatter)
+        return handler
+
+    handler = make_handler(log_dir, log_prefix, logging.DEBUG)
+
+    # Ansible ansible.builtin.script feature doesn't seperate stdout
+    # and stderr when the text is returned to the calling .yaml. Logging
+    # to stdout will cause problems with ansible checking the return
+    # strings from the python script.
+
+    # stdout_handler = logging.StreamHandler(stream=sys.stdout)
+
+    logger.addHandler(handler)
+    # logger.addHandler(stdout_handler)
+
+
+def run_command(argv, stdin=None):
+    logger.info("Running... %s", " ".join(argv))
+
+    try:
+        proc = subprocess.Popen(
+            argv, stdin=stdin, stdout=subprocess.PIPE, close_fds=True
+        )
+
+        out, err = proc.communicate()
+
+        if err is not None:
+            logger.info("Error running %s: %s", argv[0], err)
+
+        out = out.decode("utf-8")
+    except OSError as e:
+        logger.info("Error running %s: %s", argv[0], e.strerror)
+        raise
+
+    logger.info("Return code: %d", proc.returncode)
+    for line in out.splitlines():
+        logger.info("%s", line)
+
+    return (proc.returncode, out)
+
+
+def check_positive(value):
+    try:
+        value = int(value)
+        if value <= 0:
+            raise argparse.ArgumentTypeError(
+                "{0:d} is not a positive integer".format(value)
+            )
+    except ValueError as error:
+        raise Exception("{0:04x} is not an integer".format(value)) from error
```

Comment:
I've never seen this before - `raise ... from ex` - but python 2 on EL7 does not like it

Reply:
or - if it is valid for python3, and you want the script to only work with python3, then we'll need to have the role install python3 on EL7 systems